flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackendBuilder.java

```diff
@@ -0,0 +1,490 @@
+		return this;
+	}
+
+	public RocksDBKeyedStateBackendBuilder<K> setDb(RocksDB db) {
+		this.db = db;
+		return this;
+	}
+
+	private static void checkAndCreateDirectory(File directory) throws IOException {
+		if (directory.exists()) {
+			if (!directory.isDirectory()) {
+				throw new IOException("Not a directory: " + directory);
+			}
+		} else if (!directory.mkdirs()) {
+			throw new IOException(String.format("Could not create RocksDB data directory at %s.", directory));
+		}
+	}
+
+	public RocksDBKeyedStateBackend<K> build() throws IOException {
+		RocksDBKeyedStateBackend<K> backend = null;
+		RocksDBWriteBatchWrapper writeBatchWrapper = null;
+		ColumnFamilyHandle defaultColumnFamilyHandle = null;
+		RocksDBNativeMetricMonitor nativeMetricMonitor = null;
+		try {
+			prepareDirectories();
+
+			RocksDBIncrementalRestorePrepareResult restorePrepareResult = null;
+			if (restoreStateHandles != null && !restoreStateHandles.isEmpty()) {
+				restorePrepareResult = checkAndPrepareRestore();
+			}
+
+			if (db == null) {
+				db = RocksDBOperationUtils.openDB(dbPath, columnFamilyDescriptors,
+					columnFamilyHandles, columnFamilyOptions, dbOptions);
+				// remove the default column family which is located at the first index
+				defaultColumnFamilyHandle = columnFamilyHandles.remove(0);
+			}
+			writeBatchWrapper = new RocksDBWriteBatchWrapper(db, writeOptions);
+			// init snapshot strategy after db is assured to be initialized
+			initializeSnapshotStrategy();
+			if (nativeMetricOptions.isEnabled()) {
+				nativeMetricMonitor = new RocksDBNativeMetricMonitor(db, nativeMetricOptions, metricGroup);
+			}
+
+			backend = new RocksDBKeyedStateBackend<>(
+				this.operatorIdentifier,
+				this.userCodeClassLoader,
+				this.instanceBasePath,
+				this.dbOptions,
+				this.columnFamilyOptions,
+				this.kvStateRegistry,
+				this.keySerializer,
+				this.numberOfKeyGroups,
+				this.keyGroupRange,
+				this.executionConfig,
+				this.numberOfTransferingThreads,
+				this.ttlTimeProvider,
+				this.db,
+				this.kvStateInformation,
+				this.keyGroupPrefixBytes,
+				this.cancelStreamRegistry,
+				this.keyGroupCompressionDecorator,
+				this.rocksDBResourceGuard,
+				this.checkpointSnapshotStrategy,
+				this.savepointSnapshotStrategy,
+				writeBatchWrapper,
+				defaultColumnFamilyHandle,
+				nativeMetricMonitor);
+			PriorityQueueSetFactory priorityQueueFactory = initPriorityQueueFactory(backend);
+			backend.setPriorityQueueFactory(priorityQueueFactory);
+
+			if (restorePrepareResult != null) {
+				postBackendInitialization(backend, restorePrepareResult);
+			}
+
+			return backend;
+		} catch (Throwable e) {
+			// Do clean up
+			if (backend != null) {
+				backend.dispose();
+			} else {
+				IOUtils.closeQuietly(defaultColumnFamilyHandle);
+				IOUtils.closeQuietly(writeBatchWrapper);
+				IOUtils.closeQuietly(db);
+				IOUtils.closeQuietly(cancelStreamRegistry);
+			}
+			// Log and rethrow
+			if (e instanceof IOException) {
```

Comment:
We throw `IOException` because the `createKeyedStateBackend` method in `AbstractKeyedStateBackend` throws IOException. Let me introduce a `BackendBuildingException` which extends IOException.